### PR TITLE
Add import can now add project references.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide Please take a look.

I definitely recommend reading this one commit at a time.

**Comment by davkean** [comment]
All your failures are caused by: #6985. Closing and reopening.

**Comment by davkean** [comment]
Logged: https://github.com/dotnet/roslyn/issues/6991 for the linux failure.

retest prtest/lin/dbg/unit32 please

**Comment by jaredpar** [comment]
CNA?

**Comment by dpoeschl** [comment]
Left some comments on the individual commits, but the general approach seems correct to me so far. 

How are we handling undo here?

Also, can you post a screenshot here of what this will look like in the lightbulb? Might be good to get discussion on the Lightbulb text -- I'm worried about us not having an indicator in the title. It might be too difficult to distinguish between a "safe" add using and a "world altering" add using.

**Comment by dpoeschl** [comment]
@CyrusNajmabadi Done with this round of feedback. Still interested in undo & the potential screenshot-based titles discussion.

**Comment by dpoeschl** [comment]
@CyrusNajmabadi Lots of actual AddUsingTests failing.

File: src/Features/Core/Portable/CodeFixes/AddImport/AbstractAddImportCodeFixProvider.SymbolReference.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+            {
+                return INamespaceOrTypeSymbolExtensions.CompareNamespaceOrTypeSymbols(this.Symbol, other.Symbol);
+            }
+
+            public abstract Solution UpdateSolution(Document newDocument);
+        }
+
+        private class ProjectSymbolReference : SymbolReference
+        {
+            private readonly ProjectId projectId;
```
Comment (jaredpar) [inline src/Features/Core/Portable/CodeFixes/AddImport/AbstractAddImportCodeFixProvider.SymbolReference.cs]: Private field use `_` prefix

File: src/Features/Core/Portable/CodeFixes/AddImport/AbstractAddImportCodeFixProvider.SymbolReference.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+            public int CompareTo(SymbolReference other)
+            {
+                return INamespaceOrTypeSymbolExtensions.CompareNamespaceOrTypeSymbols(this.Symbol, other.Symbol);
+            }
+
+            public abstract Solution UpdateSolution(Document newDocument);
+        }
+
+        private class ProjectSymbolReference : SymbolReference
```
Comment (jaredpar) [inline src/Features/Core/Portable/CodeFixes/AddImport/AbstractAddImportCodeFixProvider.SymbolReference.cs]: sealed?

File: src/Features/Core/Portable/CodeFixes/AddImport/AbstractAddImportCodeFixProvider.SymbolReference.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+                // If this reference came from searching another project, then add a project reference
+                // as well.
+                var newProject = newDocument.Project;
+                newProject = newProject.AddProjectReference(new ProjectReference(this.projectId));
+
+                return newProject.Solution;
+            }
+        }
+
+        private class MetadataSymbolReference : SymbolReference
```
Comment (jaredpar) [inline src/Features/Core/Portable/CodeFixes/AddImport/AbstractAddImportCodeFixProvider.SymbolReference.cs]: sealed?

File: src/Features/Core/Portable/CodeFixes/AddImport/AbstractAddImportCodeFixProvider.SymbolReference.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        {
+            private readonly PortableExecutableReference reference;
```
Comment (jaredpar) [inline src/Features/Core/Portable/CodeFixes/AddImport/AbstractAddImportCodeFixProvider.SymbolReference.cs]: private field prefix with _

**Comment by CyrusNajmabadi** [comment]
@jaredpar regarding naming.  Is there just some tool i can run that will fix this all up for me in one go?  I'd like to take all the services code and bring it up to spec rather that taking a piecemeal approach.

Regarding 'sealed'.  in general, for internal APIs, i'm not one to seal things.  In the case here, sealing would mean to me "do not subclass this, that is wrong and you should feel bad".  However, in this domain, subclassing could be totally appropriate for the next extender/maintainer depending on what they were trying to accomplish.

**Comment by jaredpar** [comment]
@CyrusNajmabadi we use [codeformatter](https://github.com/dotnet/codeformatter) to enforce the naming / style guidelines.  You can run that if you like on the affected projects.

**Comment by Pilchie** [comment]
Added various comments. Generally :+1:, but I also think we should handle undo, and would like to know what the descriptions look like now (as @dpoeschl said)

**Comment by CyrusNajmabadi** [comment]
> Nothing ever passes false for this anywhere.

When we search projects that we don't have references to, we pass "false" since we don't want to search *that* project's direct references.

**Comment by CyrusNajmabadi** [comment]
> It's weird to have equality only on the base type and only look at one of the possible fields.

Except, for the purposes of "Add Import" it's all that's necessary.  You can't have symbols that are equal produced from different projects/dlls.

**Comment by CyrusNajmabadi** [comment]
Here are examples of what it looks like:

image

image

**Comment by dpoeschl** [comment]
@CyrusNajmabadi I meant the titles as they appear in the lightbulb listing *before* they're expanded. Also, what's the status of undo?

**Comment by CyrusNajmabadi** [comment]
The titles in the lightbulb are the same as they are today.  Nothing has changed there.  i.e. it will show something like "using System.Xml;".

We've started an email thread about Undo to see what is the best way to accomplish this in VS.  I'm thinking it will be necessary to create our own IOleUndoUnit for this.

**Pull request merged**